src/amici.cpp

```diff
@@ -181,436 +113,355 @@ void *setupAMI(int *status, UserData *udata, TempData *tdata) {
+            
             break;
-
-        case AMI_SPTFQMR:
-            *status = AMISptfqmr(ami_mem, PREC_NONE, 5);
-            if (*status != AMI_SUCCESS) return(NULL);
-
-            *status = wrap_SetJacTimesVecFn(ami_mem);
-            if (*status != AMI_SUCCESS) return(NULL);
-
+            
+        case AMICI_SPTFQMR:
+            if (AMISptfqmr(ami_mem, PREC_NONE, 5) != AMICI_SUCCESS) return(NULL);
+            
+            if (wrap_SetJacTimesVecFn(ami_mem) != AMICI_SUCCESS) return(NULL);
+            
             break;
-
+            
             /* SPARSE SOLVERS */
-
-        case AMI_KLU:
-            *status = AMIKLU(ami_mem, udata->nx, udata->nnz, CSC_MAT);
-            if (*status != AMI_SUCCESS) return(NULL);
-
-            *status = wrap_SetSparseJacFn(ami_mem);
-            if (*status != AMI_SUCCESS) return(NULL);
-
-            *status = AMIKLUSetOrdering(ami_mem, udata->ordering);
-            if (*status != AMI_SUCCESS) return(NULL);
-
+            
+        case AMICI_KLU:
+            if (AMIKLU(ami_mem, udata->nx, udata->nnz, CSC_MAT) != AMICI_SUCCESS) return(NULL);
+            
+            if (wrap_SetSparseJacFn(ami_mem) != AMICI_SUCCESS) return(NULL);
+            
+            if (AMIKLUSetOrdering(ami_mem, udata->ordering) != AMICI_SUCCESS) return(NULL);
+            
             break;
-
+            
         default:
             errMsgIdAndTxt("AMICI:mex:solver","Invalid choice of solver!");
             break;
     }
-
-    if (udata->sensi >= AMI_SENSI_ORDER_FIRST) {
-
-        tdata->am_dydx = new realtype[udata->ny * udata->nx]();
-        tdata->am_dydp = new realtype[udata->ny * udata->nplist]();
-        dgdp = new realtype[udata->ng * udata->nplist * udata->nytrue]();
-        dgdx = new realtype[udata->ng * udata->nxtrue * udata->nt]();
-        dgdy = new realtype[udata->nytrue * udata->ng * udata->ny]();
-        if (udata->ne > 0) {
-            dzdp = new realtype[udata->nz*udata->nplist]();
-            dzdx = new realtype[udata->nz*udata->nx]();
-        }
-        drdp = new realtype[udata->ng * udata->nplist * udata->nztrue * udata->nmaxevent]();
-        drdx = new realtype[udata->ng * udata->nx * udata->nztrue * udata->nmaxevent]();
-
-        dsigma_ydp = new realtype[udata->ny * udata->nplist]();
-        if(udata->ne>0) dsigma_zdp = new realtype[udata->nz * udata->nplist]();
-
-        if (udata->sensi_meth == AMI_SENSI_FSA) {
-
-            if(udata->nx>0) {
-
-                /* allocate some more temporary storage */
-                NVsx = N_VCloneVectorArray_Serial(udata->nplist, tdata->am_x);
-                sdx = N_VCloneVectorArray_Serial(udata->nplist, tdata->am_x);
-                if (NVsx == NULL) return(NULL);
-                if (sdx == NULL) return(NULL);
-
+    
+    if (udata->sensi >= AMICI_SENSI_ORDER_FIRST) {
+        if (udata->sensi_meth == AMICI_SENSI_FSA) {
+            if (udata->nx>0) {
```

Comment:
We could replace all these multiple checks by (not yet existing) `bool UserData::needFSA()` to increase readability.